state-chain/pallets/cf-elections/src/electoral_systems/block_height_tracking.rs

```diff
@@ -81,212 +82,364 @@ pub enum ChainProgress<ChainBlockNumber> {
+				if let Err(err) = validate_vote_and_height(*witness_from_index, &vote.0) {
 					log::warn!("received invalid vote: {err:?} ");
 					continue;
 				}
 
 				// we count a given vote as multiple votes for all nonempty subchains
-				while vote.len() > 0 {
-					consensus.insert_vote((vote.len(), vote.clone()));
-					vote.pop_back();
+				while vote.0.len() > 0 {
+					consensus.insert_vote((vote.0.len(), vote.clone()));
+					vote.0.pop_back();
 				}
 			}
 
-			Ok(consensus
-				.check_consensus(&Threshold {
-					threshold: success_threshold_from_share_count(num_authorities),
+			consensus.check_consensus(&threshold).map(|result| {
+				log::info!(
+					"(witness_from: {:?}): successful consensus for ranges: {:?}..={:?}",
+					witness_from_index,
+					result.0.front(),
+					result.0.back()
+				);
+				result
+			})
+		}
+	}
+}
+
+//------------------------ input headers ---------------------------
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Encode, Decode, TypeInfo, MaxEncodedLen)]
+pub struct InputHeaders<H, N>(pub VecDeque<Header<H, N>>);
+
+#[cfg(test)]
+mod tests {
+
+	use sp_std::collections::vec_deque::VecDeque;
+
+	use crate::electoral_systems::block_height_tracking::state_machine::{Indexed, Validate};
+	use proptest::{
+		prelude::{any, prop, Arbitrary, Just, Strategy},
+		prop_oneof, proptest,
+	};
+
+	use super::{
+		primitives::Header, state_machine::StateMachine, BHWState, BlockHeightTrackingDSM,
+		InputHeaders,
+	};
+
+	pub fn arb_input_headers<
+		H: Arbitrary + Clone,
+		N: Arbitrary + Clone + 'static + sp_std::ops::Add<N, Output = N> + From<u32>,
+	>(
+		witness_from: N,
+	) -> impl Strategy<Value = InputHeaders<H, N>> {
+		prop::collection::vec(any::<H>(), 2..10).prop_map(move |data| {
+			let headers =
+				data.iter().zip(data.iter().skip(1)).enumerate().map(|(ix, (h0, h1))| Header {
+					block_height: N::from(ix as u32) + witness_from.clone(),
+					hash: h1.clone(),
+					parent_hash: h0.clone(),
+				});
+			InputHeaders::<H, N>(headers.collect())
+		})
+
+		// (any::<H>(), any::<H>()).prop_map(move |(h, p)| {
+		// 	InputHeaders::<H, N>(VecDeque::from(vec![Header {
+		// 		block_height: witness_from.clone(),
+		// 		hash: h,
+		// 		parent_hash: p,
+		// 	}]))
+		// })
+		// (any::<N>(), any::<H>(), any::<H>()).prop_map(move |(n, h, p)|
+		// InputHeaders::<H,N>(VecDeque::from(vec![ Header { block_height: n, hash: h, parent_hash:
+		// p }])))
+	}
+
+	pub fn arb_state<
+		H: Arbitrary + Clone,
+		N: Arbitrary + Clone + 'static + sp_std::ops::Add<N, Output = N> + From<u32>,
+	>() -> impl Strategy<Value = BHWState<H, N>> {
+		prop_oneof![
+			Just(BHWState::Starting),
+			(any::<N>(), any::<bool>()).prop_flat_map(move |(n, b)| {
+				arb_input_headers(n).prop_map(move |headers| {
+					let witness_from = if b {
+						headers.0.front().unwrap().block_height.clone()
+					} else {
+						headers.0.back().unwrap().block_height.clone() + 1.into()
+					};
```

Comment:
This is the strategy to generate an arbitrary state value. The
```rust
Running { headers: VecDeque<Header<H, N>>, witness_from: N },
``` 
variant of `BHWState` contains the `witness_from` field, which describes what property the currently running election was created with. It is not an iteration behind, the votes we receive all have to start with this exact number, otherwise they are invalid.

Reply:
Yeah ok makes sense, thinking of it the other way around 👍